A tutorial to explain use of Rust-Aya for writing tracepoint programs (#157)
aya-rs/book: docs/book/programs/tracepoints.md

 Favorite `https://github.com/aya-rs/aya-template` not found in config, using it as a git repository: https://github.com/aya-rs/aya-template
 Project Name: aya-tracepoint-echo-open
 Destination: /home/nirmalya/Workspace-Rust/eBPF/my-second-ebpf/aya-tracepoint-echo-open ...
 project-name: aya-tracepoint-echo-open ...
 Generating template ...
✔  Which type of eBPF program? · tracepoint
 Which tracepoint category? (e.g sched, net etc...): syscalls
 Which tracepoint name? (e.g sched_switch, net_dev_queue): sys_enter_openat
 Moving generated files into: `<Current working directory>/aya-tracepoint-echo-open`...
 Initializing a fresh Git repository
 Done! New project created <Current working directory>/aya-tracepoint-echo-open
```

Note that the project's name is _aya-tracepoint-echo-open_, type of eBPF program is _tracepoint_ (from the menu), category is _syscalls_ and name of tracepoint is _sys_enter_openat_ .

The directory structure is similar to what is described [here](https://aya-rs.dev/book/start/#the-lifecycle-of-an-ebpf-program).

To build the application, move to the directory `aya-tracepoint-echo-open` and fire the following commands:

```bash
# First, build the application

> cargo xtask build-ebp

Reviewer: ```suggestion
cargo xtask build-ebpf
```

Author: Good catch! Embarrassed and Thanks!

# And, then run
RUST_LOG=info cargo xtask run
```

The output on the screen will be:

```bash
[2024-05-12T02:30:29Z INFO  aya_tracepoint_echo_open] Waiting for Ctrl-C...
[2024-05-12T02:30:29Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called
[2024-05-12T02:30:29Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called
[2024-05-12T02:30:29Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called
[2024-05-12T02:30:29Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called

> ....

Reviewer: ```suggestion
...
```

```

So, the program is running but it is incomplete. We don't know which files are being opened. Let's modify the code, to see the names of those files.

## Design (attempt 1)

For this demo program, we are going to rely on aya-log to print filenames from the eBPF program. 

Because at the tracepoint `sys_enter_openat`, the kernel knows the name of the file (the path to the file, including relative path), we should be able to ask kernel to share that with us. Then, we can print the complete name of the file that is being opened.

For this to happen, such an approach will work:

```rust

> // Source: aya-tracepoint-echo-open-ebpf/src/main.rs

Reviewer: Please don't include the whole code here. Let's always include the source file, like you did in line 173.

Author: Yes, I agree. I will move the whole function into the source code, name it appropriately and leave a link to it in the Markdown file.

    // Load the pointer to the filename. The offset value can be found running:
    // sudo cat /sys/kernel/debug/tracing/events/syscalls/sys_enter_open/format
    const FILENAME_OFFSET: usize = 24;
    if let Ok(filename_addr) = unsafe { ctx.read_at::<u64>(FILENAME_OFFSET) } {
       // read the filename
       let filename = unsafe {
            // Get an UTF-8 String from an array of bytes
            core::str::from_utf8_unchecked(
                    // Use the address of the kernel's string
                    // to copy its contents into the array named 'buf'  
                    match bpf_probe_read_user_str_bytes (
                        filename_addr as *const u8,
                        &mut buf,
                    ) {
                        Ok(_) =>  &buf,
                        Err(e)  => {
                            info!(&ctx, "tracepoint sys_enter_openat called buf_probe failed {}", e);
                            return Err(e);
                        }, 
                    }
            )

        };

        info!(&ctx, "tracepoint sys_enter_openat called, filename  {}", filename);  
    }
    Ok(0)
}
```

## eBPF code

> - This is the code (its skeleton is generated by `aya-tool`) that runs in the Kernel's eBPF Virtual Machine. The pattern highlights how aya programs are structured: `aya_tracepoint_echo_open(ctx: TracePointContext)` is a public function; it delegates the actual eBPF task to a another function ( `try_aya_tracepoint_echo_open(ctx: TracePointContext) -> Result<u32, i64>`).

Reviewer: ```suggestion
- This is the code (its skeleton is generated by `cargo generate`) that runs in the Kernel's eBPF Virtual Machine. The pattern highlights how aya programs are structured: `aya_tracepoint_echo_open(ctx: TracePointContext)` is a public function; it delegates the actual eBPF task to a another function ( `try_aya_tracepoint_echo_open(ctx: TracePointContext) -> Result<u32, i64>`).
```

aya-tool wasn't used here, as far as I can see.

Author: Good catch, Thanks.

- The content of what `filename_addr` is pointing to, is copied into a byte-array held inside the function, as an UTF-8 string.
- Print the UTF-8 string.

It works. When run, the output is like this:

```console
$ RUST_LOG=info cargo xtask run
[2024-05-16T15:32:29Z INFO  aya_tracepoint_echo_open] Waiting for Ctrl-C...
[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /proc/meminfo
[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
[2024-05-16T15:32:30Z INFO  aya_tracepoint_echo_open] tracepoint sys_enter_openat called, filename  /sys/fs/cgroup/
```

But the program is not designed correctly. Why? There are two reasons:

> 1.    The `buf` array is only of 16 bytes. The path-string of files being opened are likely to be much longer than this. The kernel accommodates 4096 bytes of path-string.

Reviewer: ```suggestion
1. The `buf` array is only of 16 bytes. The path-string of files being opened are likely to be much longer than this. The kernel accommodates 4096 bytes of path-string.
```

No need for so many spaces :stuck_out_tongue:

Author: Got it!

> 2.    It is not possible to hold 4096 bytes in the eBPF stack (the limit is 512 bytes). Therefore, we have to resort to some other mechanism to deal with this.

Reviewer: ```suggestion
2.  It is not possible to hold 4096 bytes in the eBPF stack (the limit is 512 bytes). Therefore, we have to resort to some other mechanism to deal with this.
```

Author: Ah, OK. Yes, too many spaces. Perhaps, I went with a <Tab>.
